core/src/main/java/com/graphhopper/routing/lm/LMRoutingAlgorithmFactory.java

```diff
@@ -0,0 +1,81 @@
+/*
+ *  Licensed to GraphHopper GmbH under one or more contributor
+ *  license agreements. See the NOTICE file distributed with this work for
+ *  additional information regarding copyright ownership.
+ *
+ *  GraphHopper GmbH licenses this file to you under the Apache License,
+ *  Version 2.0 (the "License"); you may not use this file except in
+ *  compliance with the License. You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.graphhopper.routing.lm;
+
+import com.graphhopper.routing.*;
+import com.graphhopper.routing.weighting.Weighting;
+import com.graphhopper.storage.Graph;
+import com.graphhopper.util.Parameters;
+
+public class LMRoutingAlgorithmFactory implements RoutingAlgorithmFactory {
+    private final RoutingAlgorithmFactory defaultAlgoFactory;
+    private final LandmarkStorage lms;
+    private final Weighting prepareWeighting;
+    private final int defaultActiveLandmarks;
+    private final int numBaseNodes;
+
+    public LMRoutingAlgorithmFactory(LandmarkStorage lms, int defaultActiveLandmarks, int numBaseNodes) {
+        this.defaultAlgoFactory = new RoutingAlgorithmFactorySimple();
+        this.lms = lms;
+        this.defaultActiveLandmarks = defaultActiveLandmarks;
+        this.prepareWeighting = lms.getWeighting();
+        this.numBaseNodes = numBaseNodes;
+    }
+
+    @Override
+    public RoutingAlgorithm createAlgo(Graph g, AlgorithmOptions opts) {
+        RoutingAlgorithm algo = defaultAlgoFactory.createAlgo(g, opts);
+        return getPreparedRoutingAlgorithm(g, algo, opts);
+    }
+
+    private RoutingAlgorithm getPreparedRoutingAlgorithm(Graph qGraph, RoutingAlgorithm algo, AlgorithmOptions opts) {
+        int activeLM = Math.max(1, opts.getHints().getInt(Parameters.Landmark.ACTIVE_COUNT, defaultActiveLandmarks));
+        if (algo instanceof AStar) {
+            if (!lms.isInitialized())
+                throw new IllegalStateException("Initialize landmark storage before creating algorithms");
+
+            double epsilon = opts.getHints().getDouble(Parameters.Algorithms.AStar.EPSILON, 1);
+            AStar astar = (AStar) algo;
+            astar.setApproximation(new LMApproximator(qGraph, prepareWeighting, numBaseNodes, lms, activeLM, lms.getFactor(), false).
```

Comment:
Do we really want the `prepareWeighting` (the one from `LandmarkStorage`) here or should it rather be `opts.getWeighting()`?

Reply:
At the moment we do not allow "cross weighting" query, i.e. we use the weighting that was used from preparation i.e. prepareWeighting is correct (should be identical to CH)

Reply:
Ok I see and yes its the same for CH. Its just a bit misleading because `RoutingAlgorithmFactory#createAlgo(Graph, AlgoOpts)` allows passing in a weighting and then its ignored.